dataframely/columns/array.py

```diff
@@ -0,0 +1,128 @@
+            shape: The shape of the array.
+            nullable: Whether this column may contain null values.
+            primary_key: Whether this column is part of the primary key of the schema.
+                Not yet supported for the Array type.
+            check: A custom check to run for this column. Must return a non-aggregated
+                boolean expression.
+            alias: An overwrite for this column's name which allows for using a column
+                name that is not a valid Python identifier. Especially note that setting
+                this option does _not_ allow to refer to the column with two different
+                names, the specified alias is the only valid name.
+            metadata: A dictionary of metadata to attach to the column.
+        """
+        if primary_key:
+            raise ValueError(
+                "`primary_key=True` is not yet supported for the Array type."
+            )
+
+        if inner.primary_key or (
+            isinstance(inner, Struct)
+            and any(col.primary_key for col in inner.inner.values())
+        ):
+            raise ValueError(
+                "`primary_key=True` is not yet supported for inner types of the Array type."
+            )
+
+        # We disallow validation rules on the inner type since Polars arrays currently don't support .eval(). Converting
+        # to a list and calling .list.eval() is possible, however, since the shape can have multiple axes, the recursive
+        # conversion could have significant performance impact. Hence, we simply disallow inner validation rules.
+        # Another option would be to allow validation rules only for sampling, but not enforce them.
+        if inner.validation_rules(pl.col(alias or "")):
+            raise ValueError(
+                "Validation rules on the inner type of Array are not yet supported."
+            )
+
+        super().__init__(
+            nullable=nullable,
+            primary_key=False,
+            check=check,
+            alias=alias,
+            metadata=metadata,
+        )
+        self.inner = inner
+        self.shape = self._ensure_tuple(shape)
+
+    @property
+    def dtype(self) -> pl.DataType:
+        return pl.Array(self.inner.dtype, self.shape)
+
+    @staticmethod
+    def _ensure_tuple(value: int | tuple[int, ...]) -> tuple[int, ...]:
+        if isinstance(value, tuple):
+            return value
+        return (value,)
+
+    def sqlalchemy_dtype(self, dialect: sa.Dialect) -> sa_TypeEngine:
+        # NOTE: We might want to add support for PostgreSQL's ARRAY type or use JSON in the future.
+        raise NotImplementedError("SQL column cannot have 'Array' type.")
+
+    def _pyarrow_dtype_of_shape(self, shape: Sequence[int]) -> pa.DataType:
+        if shape:
+            size, *rest = shape
+            return pa.list_(self._pyarrow_dtype_of_shape(rest), size)
+        else:
+            return self.inner.pyarrow_dtype
```

Comment:
I think we have to be more explicit about the check here as, otherwise, we'd end up in the `else` branch when the user provides an "empty shape" (i.e. `shape=0`). Could you add a test that we correctly return a list if we pass `shape=0`?

```suggestion
        if len(shape) > 0:
            size, *rest = shape
            return pa.list_(self._pyarrow_dtype_of_shape(rest), size)
        else:
            return self.inner.pyarrow_dtype
```

Reply:
The current code works because shape is a `tuple[int, ...]` and cannot be an `int`, and `bool()` of a non-emtpy tuple evaluates to `True`.

I parametrised https://github.com/Quantco/dataframely/blob/add-array-column/tests/columns/test_pyarrow.py#L53-L61 to demonstrate that this works on `shape=0`, too.

Let me know if you still prefer the explicit `len(shape) > 0`.

Reply:
> bool() of a non-emtpy tuple evaluates to True.

Ah sorry, I somehow thought that `bool((0,))` evaluates to `False` but it doesn't :smile: no need to change then!